### ISSUE TITLE
Should subs match bound dummy variables?

### ISSUE THREAD

**Issue by oscarbenjamin** [sympy/sympy#17794]
Using subs for expressions with different dummy variables means that the substitution doesn't happen:
```julia
In [19]: xi1 = Dummy('xi')

In [20]: xi2 = Dummy('xi')

In [21]: expr1 = Subs(f(xi1).diff(xi1), xi1, y)

In [22]: expr2 = Subs(f(xi2).diff(xi2), xi2, y)

In [23]: expr1
Out[23]: 
⎛d       ⎞│   
⎜──(f(ξ))⎟│   
⎝dξ      ⎠│ξ=y

In [25]: expr1.subs(expr2, 0)
Out[25]: 
⎛d       ⎞│   
⎜──(f(ξ))⎟│   
⎝dξ      ⎠│ξ=y

In [26]: expr1.doit().subs(expr2.doit(), 0)
Out[26]: 0
```
In fact we even have:
```julia
In [24]: expr1.subs(expr1, 0)
Out[24]: 
⎛d       ⎞│   
⎜──(f(ξ))⎟│   
⎝dξ      ⎠│ξ=y
```
Should subs be cleverer about bound variables when matching expressions to substitute? Or is this just another example of why it's a bad idea to substitute for expressions?

This is the cause of the following no-cache test failure:
```console
$ SYMPY_USE_CACHE="no" pytest sympy/diffgeom/tests/test_diffgeom.py -k test_covar_deriv
========================================== test session starts ==========================================
platform linux -- Python 3.6.3, pytest-5.0.1, py-1.8.0, pluggy-0.12.0
architecture: 64-bit
cache:        no
ground types: gmpy 2.0.8

rootdir: /space/enojb/sympy/sympy.git, inifile: pytest.ini
plugins: forked-1.0.2, xdist-1.29.0, instafail-0.4.1
collected 12 items / 11 deselected / 1 selected                                                         

sympy/diffgeom/tests/test_diffgeom.py F                                                           [100%]

=============================================== FAILURES ================================================
___________________________________________ test_covar_deriv ____________________________________________

    def test_covar_deriv():
        ch = metric_to_Christoffel_2nd(TP(R2.dx, R2.dx) + TP(R2.dy, R2.dy))
        cvd = BaseCovarDerivativeOp(R2_r, 0, ch)
        assert cvd(R2.x) == 1
        # This line fails if the cache is disabled:
>       assert cvd(R2.x*R2.e_x) == R2.e_x
E       assert x*Subs(Derivative(_#_0(_xi), _xi), _xi, x) + e_x == e_x
E        +  where x*Subs(Derivative(_#_0(_xi), _xi), _xi, x) + e_x = BaseCovarDerivativeOp(CoordSystem(rectangular, Patch(origin, Manifold(R**2, 2)), (x, y)), 0, [[[0, 0], [0, 0]], [[0, 0], [0, 0]]])((x * e_x))
E        +    where x = Manifold(R**2, 2).x
E        +    and   e_x = Manifold(R**2, 2).e_x
E        +  and   e_x = Manifold(R**2, 2).e_x

sympy/diffgeom/tests/test_diffgeom.py:118: AssertionError
                                            DO *NOT* COMMIT!                                            
================================ 1 failed, 11 deselected in 1.35 seconds ================================
```

**Comment by asmeurer** [comment]
It can be hard to access these variables otherwise so maybe it should. Unlike in your example they are usually created automatically by something else and not created explicitly by the user. 

Although I was playing around with substituting automatically created subs created from diff with the chain rule (like `diff(f(cos(x)), x)`), and I couldn't find any examples where the subs doesn't happen, even with the cache off. Your `expr1.subs(expr1, 0)` example gives 0 for me.

**Comment by oscarbenjamin** [comment]
> Your `expr1.subs(expr1, 0)` example gives 0 for me.

Yeah, I don't see that now. I must have gotten mixed up...

With `diff(f(cos(x)), x)` the expressions end up having the same dummy symbol somehow even if the cache is disabled:
```julia
$ SYMPY_USE_CACHE="no" isympy
IPython console for SymPy 1.5.dev (Python 3.6.3-64-bit) (ground types: gmpy, cache: off)
...

In [1]: srepr(diff(f(cos(x)), x))                                                                        
Out[1]: "Mul(Integer(-1), sin(Symbol('x')), Subs(Derivative(Function('f')(Dummy('xi_1', dummy_index=-4928339663424551515)), Tuple(Dummy('xi_1', dummy_index=-4928339663424551515), Integer(1))), Tuple(Dummy('xi_1', dummy_index=-4928339663424551515)), Tuple(cos(Symbol('x')))))"

In [2]: srepr(diff(f(cos(x)), x))                                                                        
Out[2]: "Mul(Integer(-1), sin(Symbol('x')), Subs(Derivative(Function('f')(Dummy('xi_1', dummy_index=-4928339663424551515)), Tuple(Dummy('xi_1', dummy_index=-4928339663424551515), Integer(1))), Tuple(Dummy('xi_1', dummy_index=-4928339663424551515)), Tuple(cos(Symbol('x')))))"
```

**Comment by oscarbenjamin** [comment]
The dummy symbols from differentiating are the same because of this:
https://github.com/sympy/sympy/blob/be9bba32042e0d24fc4a7474842a95fb281b794d/sympy/core/function.py#L775-L776
The `A` there is `cos(x)` so any expression that compares equal to that must have the same hash and get the same `Dummy`. Related issues #4624 #4719 #5600 #8510

**Comment by smichr** [comment]
I would recommend against letting `subs` access Dummy variables. Their value should be inconsequential for purposes other than display. `xreplace` should be able to access them, however.

**Comment by oscarbenjamin** [comment]
@smichr I'm not sure I understand what you are suggesting.

Firstly to be clear I am referring to bound symbols rather than Dummy symbols (dummy with a small 'd'). Their value is mathematically inconsequential except for display but subs treats them as different e.g.:
```julia
In [1]: Integral(x, (x, 0, 1)).subs(Integral(y, (y, 0, 1)), z)                                                                                                
Out[1]: 
1     
⌠     
⎮ x dx
⌡     
0 
```
Here the two integrals are mathematically identical but the substitution hasn't worked because they have different dummy variables. If they have the same dummy variables then the integral will be replaced by subs:
```julia
In [2]: Integral(x, (x, 0, 1)).subs(Integral(x, (x, 0, 1)), z)                                                                                                
Out[2]: z
```
I suspect that the line I linked above that explicitly creates a `Dummy` based on the object hash is an attempt to work around this limitation. It will probably work most of the time but theoretically can lead to wrong results since it overrides the guarantee that newly created `Dummy` symbols are unique (if any other part of the codebase did the same thing it could easily fail through matching hashes).

**Comment by asmeurer** [comment]
I think for things like Integral that don't use a Dummy (capital D), the current behavior is fine. My worry is with Subs that uses a Dummy (or rather, it's created with a Dummy by diff), if you can get the behavior described in the OP. But maybe the dummy_index thing that is already there takes care of the issue. 

The only problem would be if you try to create the Subs yourself manually using another Dummy, instead of using the one created for you by diff.

**Comment by smichr** [comment]
@oscarbenjamin , although I write with capital D I, too, meant small 'd'. But I see what you mean. Somewhere at sometime I had worked on a canonical method which would return an expression in canonical form in terms of bound symbols. The other alternative is that any object that has dummies and wants to allow full object matching would have to define its own `_subs` routine (or whatever is prescribed in the subs notes). It could be as simple as doing a `dummy_eq` and returning None if there is no match to allow fallback substitution to take place.

**Comment by asmeurer** [comment]
It's clearer to say "bound variable". Bound variables are the opposite of free variables. They can sometimes be set to Dummy variables, but they are not necessarily. For instance Subs doesn't always use a Dummy but diff creates one with a Dummy (because it is a new variable). IIRC the matrix expression -> summation methods also do this. The logic discussed here specifically applies to bound variables that are also Dummy objects. 

I think this logic would be the same for any expression with bound variables. You can detect it by looking for the Dummy symbols in the expression and seeing if they are not in free_symbols (I'm guessing `dummy_eq` already does this or something similar). So if we had this logic, it would go in the default subs routine instead of reimplementing it for any expression with bound variables.

**Comment by oscarbenjamin** [comment]
We can get the bound variables like so:
```julia
In [5]: e = Integral(x, (x, y, z))

In [6]: e
Out[6]: 
z     
⌠     
⎮ x dx
⌡     
y     

In [7]: e.free_symbols
Out[7]: {y, z}

In [8]: e.atoms(Symbol)
Out[8]: {x, y, z}

In [9]: e.atoms(Symbol) - e.free_symbols
Out[9]: {x}
```
A few objects define a bound_symbols attribute but it isn't universal:
```julia
In [11]: Lambda(x, x*y).bound_symbols
Out[11]: (x,)
```

@smichr I guess you are referring to the `as_dummy` method which puts an object into canonical form wrt bound variables (although see #17498)

**Comment by smichr** [comment]
I left a comment at #17498

**Comment by smichr** [comment]
So `==` should basically never make allowances for differences in bound symbols. Making `dummy_eq` or alternate more robust, then, and using that when one wants to know if two objects are mathematically the same in terms of structure (sans differences in bound variables) would be the way to go.

**Comment by oscarbenjamin** [comment]
Yes and also replace should be more robust - it probably needs to be rewritten somehow.

**Comment by m93a** [comment]
I'm not sure I follow all the comments… The current consensus is that not only `subs` should allow the substitution of bound Dummy variables, but also `==` should mark as equal two expressions with different dummies but otherwise identical. Is that correct?

I would really appreciate at least the `subs` support: I want to analyze the properties of the vector field `f(x**2 + y**2 + z**2) * (i+j+k)`, where `f` is an arbitrary function. Both rotation and divergence of this field produce ugly expressions that are hard to read and impossible to work with.

**Minimal example:**
```python
from sympy import *
from sympy.abc import a, b
wild = Wild('*')

f  = Function("f")
f_ = Function("f'")

expr = Derivative(f(a+b), a).doit()

expr.subs( f(wild).diff(wild), f_ )
# nope, still the same
```
![image](https://user-images.githubusercontent.com/1671665/73598387-043fda00-4538-11ea-9a91-3725f70d8e78.png)


**Actual usage example:**
```python
import sympy
from sympy import *
from sympy.vector import *

cart = vector.CoordSys3D('')
x, y, z = cart.base_scalars()
i, j, k = cart.base_vectors()

f = Function('f')
F = f(x**2 + y**2 + z**2) * (i+j+k)
F.to_matrix(cart)
```
![image](https://user-images.githubusercontent.com/1671665/73598413-5254dd80-4538-11ea-8c52-732ad8bafd67.png)
```python
crl = Curl(F).doit()

f_ = Function('f\'')
wild = Wild('*')

crl.subs(f(wild).diff(wild), f_).to_matrix(cart)
```
![image](https://user-images.githubusercontent.com/1671665/73598427-86c89980-4538-11ea-9c7f-63d46f4ece44.png)

Not only that I can barely read the expression, but if I wanted to set `f'` to an actual function, I can't. The desired output would of course be _this expression_, where I could factor out `2f'`, and substitute `f'` for eg. `sqrt` or something:
![image](https://user-images.githubusercontent.com/1671665/73598506-6816d280-4539-11ea-8122-8fa6d58f9519.png)

**Comment by oscarbenjamin** [comment]
> I'm not sure I follow all the comments… The current consensus is that not only `subs` should allow the substitution of bound Dummy variables, but also `==` should mark as equal two expressions with different dummies but otherwise identical. Is that correct?

No, I don't think so. I think that the consensus is:

The `==` operator is for structural equality. Expressions with different symbols should not compare equal unless the symbols do. The difference between a `Dummy` and a normal `Symbol` is just that a `Dummy` is not considered equal to another `Dummy` just because they might happen to have the same name. Also `==` should not do anything clever for bound symbols.

However `subs` is a high-level mathematical operation that does understand bound symbols and could do something clever about `Dummy` bound symbols when checking expressions for equality. It should not use `==` for this but something like `dummy_eq` instead though.

**Comment by m93a** [comment]
Okay, thanks for clarification! After the exam period I might look into it and see if I can do a PR.
Btw, do you understand why the `Dummy` variables don't get matched even by `Wild`? Seems quite counter-intuitive to me.

**Comment by oscarbenjamin** [comment]
I think that they should be matched by `Wild`. Can you open a separate issue for that?

**Comment by m93a** [comment]
Oops, my bad.
I was using `subs` instead of `replace`. Using `replace` with `Wild` works just fine 🙂

**Comment by smichr** [comment]
> We can get the bound variables like so:

Yes, that is right, but since bound symbols need not be different there is no way to tell which one goes with which part of the expression (and hence, `subs` can be ambiguous):
```python
(Integral(x,(x,1,2)) + Sum(x, (x, (1,3)))).subs(x,y)  # which one?, the Integral or Sum
```
An unambiguous way to do subs of these bound symbols is to convert the symbol in the part of the expression that has the bound symbol of interest, i.e. replace the Sum with an ".xreplaced x" form. And the set of sub-expressions with bound symbols would be
```python
set([i for i in expr.atoms(Basic) if i.atoms(Symbol) - i.free_symbols])
```
But even there there may be ambiguity since you could have identical structures in two different places in the expression. Only the unique sub-expression containing the object of interest can disambiguate the intent of the subs.

For these complicating reasons, I would discourage modifying `subs` to allow for this. I wonder if any of the not-oft-discussed epath-related functions could handle this.
